File: sort-algo/sort_algo/merge_sort.py
def merge_sort(l: list[int]) -> list[int]:
    if l == []:
        return []
    if len(l) == 1:
        return l[:]
    mid = len(l) // 2
    l1 = l[:mid]
    l2 = l[mid:]
    l1_sorted = merge_sort(l1)
    l2_sorted = merge_sort(l2)
    l_sorted = merge(l1_sorted, l2_sorted)
    return l_sorted


def merge(l1: list[int], l2: list[int]):
    i1 = 0
    i2 = 0

    merged: list[int] = []

    while i1 < len(l1) and i2 < len(l2):
        if l1[i1] < l2[i2]:
            merged.append(l1[i1])
            i1 += 1
        else:
            merged.append(l2[i2])
            i2 += 1

    if i1 == len(l1):
        merged += l2[i2:]
    else:
        merged += l1[i1:]

    return merged

File: sort-algo-ts/merge_sort.ts
export function merge_sort(l: number[]): number[] {
  if (l.length === 0) return l.slice();
  if (l.length === 1) return l.slice();
  const mid = Math.floor(l.length / 2);
  const left = l.slice(0, mid);
  const right = l.slice(mid);
  const left_sorted = merge_sort(left);
  const right_sorted = merge_sort(right);
  return merge(left_sorted, right_sorted);
}

export function merge(l1: number[], l2: number[]) {
  let i1 = 0;
  let i2 = 0;
  let result: number[] = [];
  while (i1 < l1.length && i2 < l2.length) {
    if (l1[i1] < l2[i2]) {
      result.push(l1[i1]);
      i1++;
    } else {
      result.push(l2[i2]);
      i2++;
    }
  }
  let l1_leftover = l1.slice(i1);
  let l2_leftover = l2.slice(i2);
  result.push(...l1_leftover);
  result.push(...l2_leftover);
  return result;
}
